### PR TITLE
Allow to replace backslash in usernames during first login

### PR DESCRIPTION
**Pull request by skabashnyuk**
Signed-off-by: Sergii Kabashniuk <skabashniuk@redhat.com>

### What does this PR do?

Allow to replace backslash in usernames during first login
### What issues does this PR fix or reference?

https://github.com/eclipse/che/issues/18404

### How to test this PR?
1. Deploy che without OpenShift OAuth integration `chectl server:start --installer=operator   --platform=openshift  -n che --che-operator-cr-patch-yaml cr.yaml` 
```yaml
spec:
  auth:
    openShiftoAuth: false
```
2. Enable user self-registration in Che Keycloak realm
Знімок екрана 2020-11-23 о 16 23 26

3. Deploy che-server image `quay.io/skabashn/che-server:che18404`
4. Register a new user with a backslash.
Знімок екрана 2020-11-23 о 16 18 00
4. Ensure that username in che doesn't contains backslash.
Знімок екрана 2020-11-23 о 16 18 27

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [x] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [x] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [x] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [x] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [x] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [x] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/multiuser.properties [hunk 1]
```diff
@@ -185,3 +185,8 @@ che.keycloak.admin_username=NULL
 # from Che database. Make sense only in case ${che.keycloak.cascade_user_removal_enabled}
 # set to 'true'
 che.keycloak.admin_password=NULL
+
+# A set of characters that are going to be used to replace `\` backslash in user name during the first login.
+# Replaced user name would be stored in Che database only. It's recommended to use DNS-compliant values.
+# Value `NULL` is turning off this feature.
+che.keycloak.username.replace_backslash_in_username_with=-
```
Comment (sparkoo) [inline assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/multiuser.properties]: might we need to replace another characters as well? I'm thinking about having it more general, and having here pairs of `[search character]:[replace with]`.

Comment (skabashnyuk) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/multiuser.properties]: done

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/multiuser.properties [hunk 2]
```diff
@@ -185,3 +185,12 @@ che.keycloak.admin_username=NULL
...
 che.keycloak.admin_password=NULL
+
+# User name adjustment configuration. Represented by comma-separated key-value pairs. That is sequentially used
+# as arguments String.replaceAll function against original username.  Where 'key' is a regular expression.
+# Can be used to replace undesired characters like '\' when the user name is used as a part of k8s object for example.
+# Replaced user name would be stored in Che database only. It's recommended to use DNS-compliant values.
+# Value `NULL` is turning off this feature.
+# Example: '\\\\=-,@=-at-' This configuration will replace `\` to `-` and `@` to `-at-`. So `org\user@com` will looks
+# like `org-user-at-com`
+che.keycloak.username.replace_backslash_in_username_with=NULL
```
Comment (mshaposhnik) [inline assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/multiuser.properties]: So maybe it's somthing like `che.keycloak.username.symbols_replace_patterns` now ?

**Review by mshaposhnik — APPROVED** [review]

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/multiuser.properties (hunk 2 above)
Comment (skabashnyuk) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/multiuser.properties]: yeh. you are right.

Comment (skabashnyuk) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/multiuser.properties]: `che.keycloak.username.replacer_patterns` ?

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/multiuser.properties [hunk 3]
```diff
@@ -185,3 +185,12 @@ che.keycloak.admin_username=NULL
...
 che.keycloak.admin_password=NULL
+
+# User name adjustment configuration. Represented by comma-separated key-value pairs. That is sequentially used
+# as arguments String.replaceAll function against original username.  Where 'key' is a regular expression.
+# Can be used to replace undesired characters like '\' when the user name is used as a part of k8s object for example.
+# Replaced user name would be stored in Che database only. It's recommended to use DNS-compliant values.
+# Value `NULL` is turning off this feature.
+# Example: '\\\\=-,@=-at-' This configuration will replace `\` to `-` and `@` to `-at-`. So `org\user@com` will looks
+# like `org-user-at-com`
+che.keycloak.username.replacer_patterns=NULL
```
Comment (metlos) [inline assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/multiuser.properties]: `che.keycloak.username.replacement_patterns` sounds slightly better to me

File: multiuser/keycloak/che-multiuser-keycloak-server/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakEnvironmentInitializationFilter.java [hunk 4]
```diff
@@ -69,13 +74,18 @@ public KeycloakEnvironmentInitializationFilter(
...
-      KeycloakSettings settings) {
+      KeycloakSettings settings,
+      @Nullable @Named("che.keycloak.username.replacer_patterns") String userNameReplacerPatterns) {
     super(sessionStore, tokenExtractor);
     this.jwtParser = jwtParser;
     this.userManager = userManager;
     this.keycloakProfileRetriever = keycloakProfileRetriever;
     this.permissionChecker = permissionChecker;
     this.keycloakSettings = settings;
+    this.userNameReplacerPatterns =
```
Comment (metlos) [inline multiuser/keycloak/che-multiuser-keycloak-server/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakEnvironmentInitializationFilter.java]: `userNameReplacementPatterns` sounds ever so slightly better to me

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/multiuser.properties [hunk 5]
```diff
@@ -185,3 +185,12 @@ che.keycloak.admin_username=NULL
 # from Che database. Make sense only in case ${che.keycloak.cascade_user_removal_enabled}
 # set to 'true'
 che.keycloak.admin_password=NULL
+
+# User name adjustment configuration. Represented by comma-separated key-value pairs. That is sequentially used
```
Comment (metlos) [inline assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/multiuser.properties]: User name adjustment configuration. Che needs to use the usernames as part of K8s object names and labels and therefore has stricter requirements on their format than the identity providers usually allow (it needs them to be DNS-compliant). The adjustment is represented by comma-separated key-value pairs. These are sequentially used as arguments to the String.replaceAll function on the original username. The keys are regular expressions, values are replacement strings that replace the characters in the username that match the regular expression.
The modified username will only be stored in the Che database and will not be advertised back to the identity provider. It is recommended to use DNS-compliant characters as replacement strings (values in the key-value pairs). Set the value of this prooperty to `NULL` to turn this feature off.
Example: `\\\\=-,@=-at-` changes `\` to `-` and `@` to `-at-` so the username `org\user@com` becomes `org-user-at-com`.

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/multiuser.properties [hunk 6]
```diff
@@ -185,3 +185,12 @@ che.keycloak.admin_username=NULL
...
+# as arguments String.replaceAll function against original username.  Where 'key' is a regular expression.
+# Can be used to replace undesired characters like '\' when the user name is used as a part of k8s object for example.
+# Replaced user name would be stored in Che database only. It's recommended to use DNS-compliant values.
+# Value `NULL` is turning off this feature.
+# Example: '\\\\=-,@=-at-' This configuration will replace `\` to `-` and `@` to `-at-`. So `org\user@com` will looks
```
Comment (metlos) [inline assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/multiuser.properties]: Is the example actually correct? Wouldn't it have to be `\\=-,@=-at-` when written in the property file as opposed to the java string?

**Review by metlos — COMMENTED** [review]
Only wording suggestions, otherwise LGTM.

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/multiuser.properties (hunk 6 above)
Comment (skabashnyuk) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/multiuser.properties]: you are right. Fixed.

**Review by sparkoo — APPROVED** [review]

**Comment by skabashnyuk** [comment]
[crw-ci-test]

**Comment by skabashnyuk** [comment]
[crw-build]

**Comment by skabashnyuk** [comment]
[ci-test]

**Comment by skabashnyuk** [comment]
[ci-build]

**Comment by Katka92** [comment]
[crw-ci-test]

**Pull request merged**